### PR TITLE
Refactor namespaces to avoid collisions.

### PR DESCRIPTION
**Pull request by JimBobSquarePants**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description
This PR refactors the entire processing pipeline to a new namespace structure fixing #401 It's huge but most of the changes are simply namespace related.

Processing extension methods and individual processors now follow a pattern that groups the API by functionality.

For example processing methods pertaining to transforms adhere to the following structure.

- `SixLabors.ImageSharp.Processing.Transforms`
- `SixLabors.ImageSharp.Processing.Transforms.Processors`

Base classes and interfaces follow that same pattern by design.

Processors are internal (for now) but I would like to make them public. I can do this as part of this PR if everyone agrees.

Additionally the following optimizations and changes have been made.
- Replaced `FastArray2D<T>` with faster `DenseMatrix<T>` which now matches 1D array for speed.
- Refactored all the processors to be immutable removing initialization code from extension methods
- Renamed `QuantizedImage` to `QuantizedFrame`
- Add new `QuantizeProcessor<TPixel>` processor
- Refactored quantizers to remove per pixel virtual call
- Moved `Rgba32` where it belongs with the other pixel formats
- Moved `Bounds()` and `Size` to `IImage`.
- Fixed a bug in `DrawImage` where the default Graphics options were being overwritten.

Drawing methods follow the same pattern. The main namespaces used are `Overlays` and `Text`.

### CODE REVIEW

File: src/ImageSharp/IImage.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    /// <summary>
+    /// Encapsulates the properties and methods that describe an image.
+    /// </summary>
+    public interface IImage : IImageInfo, IDisposable
+    {
+        /// <summary>
+        /// Gets the size of the image.
+        /// </summary>
+        /// <returns>The <see cref="Size"/></returns>
+        Size Size();
```
Comment (antonfirsov) [inline src/ImageSharp/IImage.cs]: We can push this up to `IImageInfo`.
Actually there are Width + Height properties here, so maybe it should be an extension method.

File: src/ImageSharp/IImage.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+        /// <summary>
+        /// Gets the bounds of the image.
```
Comment (antonfirsov) [inline src/ImageSharp/IImage.cs]: `Bounds()`  could be an extension method on `IImageInfo`.

Comment (JimBobSquarePants) [reply src/ImageSharp/IImage.cs]: I'm happy to move them there. 👍

**Comment by antonfirsov** [comment]
Another counter-argument against publishing the processors (and I promise, this would be the last one 😄):

Implementing #142 won't be possible without breaking public API. Are we fine with this?
It would also affect `IResampler` API.

**Comment by JimBobSquarePants** [comment]
@antonfirsov You've got me there! 😄  

We can't afford another round of massive churn after this before V1 so yeah, if we want to implement #142 we should keep the processors internal for now.

I imagine value type kernels would be the last big task before V1. Do you think you would be able to allocate (haha) time in order to take that on?

**Comment by antonfirsov** [comment]
Unfortunately, I can't make for 1.0 😞, it's a quite a lot of work & I'm starting my new job in a week.

We need to decide if we should make it possible to implement it in a `1.*`  version, without major braking changes. 
The price is: keeping things internal and/or investing extra effort into making our API more future-proof.

**Comment by JimBobSquarePants** [comment]
>We need to decide if we should make it possible to implement it in a 1.* version, without major breaking changes

It all depends on how much it will affect the `IResampler` API. We've got calling overloads in `Resize` that use the type as a parameter. `Resize` will probably be our most utilized method so we need to be very careful.

I'm happy to wait til V2 for #142 also though. 

Ship a working version, hopefully make enough money to allow us all to live of this should we choose to and then add everything we've learned from V1* to V2

**Comment by antonfirsov** [comment]
Naive apporach:
`IResampler` should be something like `IResampler<TMatrix>` & `ResizeProcessor` should be also generic to utilize JIT-time, boxing & virtual call free optimizations.

Advanced approach:
Finding a way to hide generics behind nested implementation classes using tricks like double dispatch. The public API should be still touched, but in a less intrusive way.

**Comment by JimBobSquarePants** [comment]
@dlemstra Yeah much better I think. What's your thoughts on the rest of the PR?

**Comment by antonfirsov** [comment]
Had another look, and still looks good to me.

I think the only non-trivial decision here is to have the drawing code in `SixLabors.ImageSharp.Processing.Overlays` instead of `SixLabors.ImageSharp.Processing.Drawing`. Can't decide if it's good or not. Thoughts?

**Comment by JimBobSquarePants** [comment]
I've moved `Overlays` back to `Drawing` again. I think devs will expect `Drawing`

**Pull request merged**